File: modules/soli-solver/ios/SoliSolverModule.swift
import ExpoModulesCore

// Declarations for the two C symbols exported by the vendored Rust static
// library (SoliSolver.xcframework). @_silgen_name instead of importing the
// bundled C header: a Swift-only CocoaPods pod has no bridging header, and
// exposing a vendored framework's header through a module map/umbrella is the
// fragile part of pod setups — a direct symbol reference cannot break at the
// header-search-path level. The signatures MUST stay in sync with
// rust/soli-solver-ffi/include/soli_solver.h (the contract reference).
@_silgen_name("soli_solver_solve")
private func soli_solver_solve(
  _ requestJson: UnsafePointer<CChar>?
) -> UnsafeMutablePointer<CChar>?

@_silgen_name("soli_solver_free_string")
private func soli_solver_free_string(_ ptr: UnsafeMutablePointer<CChar>?)

public class SoliSolverModule: Module {
  public func definition() -> ModuleDefinition {
    Name("SoliSolver")

    // Plain-closure AsyncFunction: runs on the shared serial
    // expo.modules.AsyncFunctionQueue, off the JS thread. Trade-off: a solve
    // blocks OTHER Expo modules' async functions on that shared queue for its
    // duration — accepted because the Rust side enforces the time budget
    // internally (~1.5 s worst case; see rust-mobile.md guide).
    AsyncFunction("solvePosition") { (requestJson: String) -> String in
      // The returned pointer is Rust-heap-allocated (independent of the
      // temporary C string), so it may escape withCString.
      guard let raw = requestJson.withCString({ soli_solver_solve($0) }) else {
        // Null only on allocation failure inside Rust (see header contract).
        return "{\"status\":\"error\",\"message\":\"solver returned null\"}"
      }
      defer { soli_solver_free_string(raw) }
      return String(cString: raw)
    }
  }
}

File: modules/soli-solver/android/src/main/java/expo/modules/solisolver/SoliSolverModule.kt
package expo.modules.solisolver

import expo.modules.kotlin.functions.Coroutine
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

// JNI bridge to the Rust solver (rust/soli-solver-ffi). Package + object +
// function name together form the exported JNI symbol and must stay in sync
// with rust/soli-solver-ffi/src/android_jni.rs:
//   Java_expo_modules_solisolver_SoliSolverJni_nativeSolve
object SoliSolverJni {
  init {
    // libsoli_solver_ffi.so from android/src/main/jniLibs/<abi>/ (built by
    // scripts/build-rust-solver.sh).
    System.loadLibrary("soli_solver_ffi")
  }

  external fun nativeSolve(request: String): String
}

class SoliSolverModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("SoliSolver")

    // Coroutine bodies start on the shared expo.modules.AsyncFunctionQueue
    // HandlerThread; hop to Dispatchers.Default because a solve can block up
    // to its budget (~1.5 s, enforced inside Rust) and must not stall other
    // modules' async calls on that shared queue (see rust-mobile.md guide).
    AsyncFunction("solvePosition") Coroutine { requestJson: String ->
      withContext(Dispatchers.Default) { SoliSolverJni.nativeSolve(requestJson) }
    }
  }
}
